lib/charms/vault_k8s/v0/vault_tls.py

```diff
@@ -217,39 +246,46 @@ def _should_request_new_certificate(self) -> bool:
         if not self.charm.model.relations.get(TLS_CERTIFICATE_ACCESS_RELATION_NAME):
             return False
 
-        csr_is_in_workload = (existing_csr := self.pull_tls_file_from_workload(File.CSR))
-        fulfilled_csrs = self.tls_access.get_certificate_signing_requests(fulfilled_only=True)
+        if not (existing_csr := self.pull_tls_file_from_workload(File.CSR)):
+            return True
+
         pending_csrs = self.tls_access.get_certificate_signing_requests(unfulfilled_only=True)
+        existing_csr_is_pending = any([existing_csr in csr_obj.csr for csr_obj in pending_csrs])
+        if existing_csr_is_pending:
+            return False
+
+        fulfilled_csrs = self.tls_access.get_certificate_signing_requests(fulfilled_only=True)
+        existing_csr_is_fulfilled = any(
+            [existing_csr in csr_obj.csr for csr_obj in fulfilled_csrs]
+        )
+
         expired_certs = self.tls_access.get_expiring_certificates()
+        existing_csr_expiring = any(
+            [existing_csr in cert_obj.certificate for cert_obj in expired_certs]
+        )
 
-        existing_csr_is_fulfilled = any([existing_csr in csr.values() for csr in fulfilled_csrs])
-        existing_csr_is_pending = any([existing_csr in csr.values() for csr in pending_csrs])
-        existing_csr_expiring = any([existing_csr in cert.values() for cert in expired_certs])
+        if existing_csr_is_fulfilled and not existing_csr_expiring:
+            return False
 
-        if csr_is_in_workload:
-            if existing_csr_is_fulfilled and not existing_csr_expiring:
-                return False
-            if existing_csr_is_pending:
-                return False
         return True
```

Comment:
I think grouping up the if statements at the end makes this more readable than how it is here personally, but it looks like the logic is still the same so it's not that important.

Reply:
I agree, it was slightly more readable
but here we return early if it's the case, and readability has not gone that bad.
I don't insist on keeping it tho
let me know

Reply:
I don't see not returning early here as too much of a big deal since we only expect vault to have a single CSR in this relation databag. I'd prefer to revert it.

Reply:
Alright, done.